src/Rings/mpoly-graded.jl

```diff
@@ -1346,61 +1346,103 @@ function homogeneous_component(R::MPolyDecRing, g::IntegerUnion)
   return homogeneous_component(R, grading_group(R)([g]))
 end
 
-function vector_space(K::AbstractAlgebra.Field, e::Vector{T}; target = nothing) where {T <:MPolyRingElem}
+
+@doc raw"""
+  vector_space(K::Field, polys::Vector{T}; target = nothing) where {T <:M PolyRingElem}
+
+Return a K-vector space `V` and an epimorphism from `V` onto the K-vector
+space spanned by the polynomials in `polys`.
+
+Note that all polynomials must have the same parent `R`, and `K` must be equal
+to `base_ring(R)`. The optional keyword argument `target` can be used to
+specify `R` when `polys` is empty.
+
+```jldoctest
+julia> R, (x, y, z) = polynomial_ring(QQ, ["x", "y", "z"]);
+
+julia> polys = [x + y, x - y, 2x + 3y];
+
+julia> V, VtoPoly = vector_space(QQ, polys)
+(Vector space of dimension 3 over QQ, Map: V -> R)
+
+julia> VtoPoly.(basis(V))
+3-element Vector{QQMPolyRingElem}:
+ x
+ y
+ 0
+```
+"""
+function vector_space(K::Field, polys::Vector{T}; target = nothing) where {T <: MPolyRingElem}
   local R
-  if length(e) == 0
+  if length(polys) == 0
     R = target
     @assert R !== nothing
   else
-    R = parent(e[1])
+    R = parent(polys[1])
+    @assert target === nothing || target === R
   end
   @assert base_ring(R) == K
-  mon = Dict{elem_type(R), Int}()
-  mon_idx = Vector{elem_type(R)}()
+  expvec = Dict{Vector{Int}, Int}()
+  expvec_idx = Vector{Vector{Int}}()
   M = sparse_matrix(K)
-  last_pos = 1
-  for i = e
+  
+  # take polynomials and turn them into sparse row vectors
+  for f in polys
     pos = Vector{Int}()
     val = Vector{elem_type(K)}()
-    for (c, m) = Base.Iterators.zip(AbstractAlgebra.coefficients(i), AbstractAlgebra.monomials(i))
-      if haskey(mon, m)
-        push!(pos, mon[m])
-        push!(val, c)
-      else
-        push!(mon_idx, m)
-        mon[m] = last_pos
-        push!(pos, last_pos)
-        push!(val, c)
-        last_pos += 1
+    for (c, e) = Base.Iterators.zip(AbstractAlgebra.coefficients(f), AbstractAlgebra.exponent_vectors(f))
+      i = get!(expvec, e) do
+        push!(expvec_idx, e)
+        length(expvec_idx)
       end
+      push!(pos, i)
+      push!(val, c)
     end
     push!(M, sparse_row(K, pos, val))
   end
+
+  # row reduce
   rref!(M)
 
+  # turn the reduced sparse rows back into polynomials
   b = Vector{elem_type(R)}()
-  for i=1:nrows(M)
-    s = zero(e[1])
-    for (k,v) = M[i]
-      s += v*mon_idx[k]
+  for i in 1:nrows(M)
```

Comment:
Is there a reason to not just go till `rank(M)` (which `rref!` returns)? Or is it necessary that the dimension of the vector space is the number of polynomials?

Reply:
I tried to not change the semantics of this function, and if I understand you right, that change would (and then also the docstring would have to be updated).

What @Lax202 really needs the code for is to get the space of all linear combinations of the given polynomials that equal zero, i.e. the kernel of the morphismus return here (only we can't get it directly, we have to jump through some hoops). I believe the work required for that is identical in either case, so from that perspective I wouldn't mind changing to `rank(M)` here. But other code might be affected, and so while I generally support the idea, I wouldn't want to try it in this PR, as it might be opening a can of worms...

Reply:
I have no stakes in this. I'm just wondering: For example, in the doctest the vector space is of dimension 3, but the last basis element maps to 0, so why not make it dimension 2?
That it always has been like this, is of course a good argument :)

(I can also try it out once this pull request is merged and see how much breaks.)